### DISCUSSION TITLE
Bug: OneToMany OrphanRemoval Not Working as Expected

### DISCUSSION THREAD

**Discussion by totalard** [mikro-orm/mikro-orm#3111] [Q&A]
**Describe the bug**
I am using MongoDB 5 in dev, & 4 in Production, 
in both , i can see orphanremoval OneToMany Not Working at all,

```
[query] db.getCollection('product').find({ '$and': [ { deletedAt: null }, { _id: ObjectId('627df318cd4293cff0c71d75') } ] }, { session: undefined }).limit(1).toArray(); [took 3 ms]
DefaultLogger.js:31 [query] db.getCollection('media').find({ _id: { '$in': [ ObjectId('627b33c35a601e6c8ba83ec3') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 2 ms]
DefaultLogger.js:31 [query] db.getCollection('product').deleteMany({ _id: { '$in': [ ObjectId('627df318cd4293cff0c71d75') ] } }, { session: undefined }); [took 2 ms]
DefaultLogger.js:31 [query] db.getCollection('product').find({ deletedAt: null }, { session: undefined }).toArray(); [took 1 ms]
DefaultLogger.js:31 [query] db.getCollection('category').find({ _id: { '$in': [ ObjectId('626e2c832794d1257d049265') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 2 ms]
DefaultLogger.js:31 [query] db.getCollection('media').find({ _id: { '$in': [ ObjectId('627b33c35a601e6c8ba83eca') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 1 ms]
DefaultLogger.js:31 [query] db.getCollection('media').find({ _id: { '$in': [ ObjectId('627fb6e0e53e2fbd683f7ff3') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 2 ms]
DefaultLogger.js:31 [query] db.getCollection('tag').find({ _id: { '$in': [ ObjectId('62661fa6f64023cfa96aed9a'), ObjectId('626620f5f64023cfa96aed9b'), ObjectId('626e23002c75772ff7e175cc') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 2 ms]
DefaultLogger.js:31 [query] db.getCollection('media').find({ _id: { '$in': [ ObjectId('627b33c35a601e6c8ba83ec9'), ObjectId('627b33c35a601e6c8ba83ec7'), ObjectId('627b33c35a601e6c8ba83ec4') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 2 ms]
DefaultLogger.js:31 [query] db.getCollection('media').find({ _id: { '$in': [ ObjectId('627b04d9794d1dc0e723c590') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 2 ms]
DefaultLogger.js:31 [query] db.getCollection('media').find({ _id: { '$in': [ ObjectId('627b33c35a601e6c8ba83ec5'), ObjectId('627b33c35a601e6c8ba83ec6'), ObjectId('627b33c35a601e6c8ba83ec7') ] } }, { session: undefined }).sort([ [ '_id', 1 ] ]).toArray(); [took 1 ms]
```

Only Delete of Product Entity is done, Do Further Delete Query Called At all

**To Reproduce**

```
@Entity()
export class Product extends BaseWithSlugSoft {
  @Property({ nullable: true })
  discription: string;

  @Property({ nullable: true })
  short: string;

  @Property({ nullable: true })
  publicNote: string;

  @Property({ unique: true })
  sku?: string = randomString(16);

  @Property({ type: t.float, default: 0.0 })
  regPrice: number;

  @Property({ type: t.float, default: 0.0 })
  salePrice: number;

  @Property({ type: t.float, default: 0.0 })
  stock: number;

  @Property()
  status: string = 'draft';

  @Property()
  isFeatured: boolean;

  @Property()
  meta: object = {};

  @OneToMany(() => Review, (review) => review.product, { orphanRemoval: true })
  reviews = new Collection<Review>(this);

  @Property({ type: t.float, default: 0.0 })
  weight: number;

  @Property({ type: t.float, default: 0.0 })
  length: number;

  @Property({ type: t.float, default: 0.0 })
  hight: number;

  @Property({ type: t.float, default: 0.0 })
  width: number;

  @Property()
  type: string;

  @ManyToMany(() => Tag, (tags) => tags.products, { owner: true })
  tags = new Collection<Tag>(this);

  @OneToMany(() => ProductAttribute, (attr) => attr.product, {
    orphanRemoval: true,
  })
  attributes = new Collection<ProductAttribute>(this);

  @ManyToMany(() => Category, (categories) => categories.products, {
    owner: true,
  })
  categories = new Collection<Category>(this);

  @ManyToMany({ entity: () => Media, mappedBy: 'products', owner: true })
  images = new Collection<Media>(this);

  @ManyToOne(() => Media, { nullable: true, eager: true })
  image?: Media;

  @ManyToMany(() => Product)
  upSell = new Collection<Product>(this);

  @ManyToMany(() => Product)
  crossSell = new Collection<Product>(this);

  @OneToMany(() => ProductVariation, (variation) => variation.product, {
    orphanRemoval: true,
  })
  varProducts = new Collection<ProductVariation>(this);
}
```

```
@Entity()
export class ProductVariation extends BaseSoftEntity {
  @Property({ unique: true })
  sku?: string = randomString(16);

  @Property({ type: t.float })
  regPrice!: number;

  @Property({ type: t.float, nullable: true })
  salePrice!: number;

  @Property({ type: t.float, default: 0 })
  stock?: number;

  @Property()
  meta?: object = {};

  @Property({ type: t.float, default: 0.0 })
  weight!: number;

  @Property({ type: t.float, default: 0.0 })
  length!: number;

  @Property({ type: t.float, default: 0.0 })
  hight!: number;

  @Property({ type: t.float, default: 0.0 })
  width!: number;

  @ManyToMany({ entity: () => Media, mappedBy: 'varProducts', eager: true })
  images = new Collection<Media>(this);

  @ManyToOne(() => Media, { nullable: true, eager: true })
  image?: Media;

  @ManyToOne(() => Product, { nullable: false, eager: true })
  product!: Product;

  @ManyToMany(() => AttributeTerm)
  terms: Collection<AttributeTerm> = new Collection<AttributeTerm>(this);

  @Property({ name: '_name' })
  getModelName() {
    return this.terms
      .getItems()
      .map((each) => each.name)
      .join(',');
  }
}
```

**Expected behavior**
It Must Call Further delate call to removeall Orphans, in OneToMany

**Additional context**
it also not working in softDelete nor normal delete

**Versions**

| Dependency | Version |
| - | - |
| node | 14 |
| mikro-orm | 5.3.1 |
| your-driver | mongo |

**Comment by totalard** [comment]
I found that , The OrphanRemoval Only worked When eager:true,

which seems not a perfect implementation, as orphanRemoval is related to **Data Integrity** in Database, eager is just prepopulating, both are eactly different, i think, orphan removal should work even there is no eager:true is set,

**Comment by B4nan** [comment]
Just a guess, but your collection needs to be initialized for orpah removal to work. 

Moving to discussions as you haven't provided any reproduction - entity definition is not enough, not talking about it being incomplete and far away from minimal on its own. If you want more help, please provide [complete but minimal reproduction](https://stackoverflow.com/help/minimal-reproducible-example). 

FYI orphan removal _is_ working just fine, mongo driver included. Tests are here: 

https://github.com/mikro-orm/mikro-orm/blob/master/tests/EntityManager.mongo.test.ts#L1627

**Reply by totalard** [reply]
But I using Softdelete as per your tweet, But i getting issue that Now OrphanRemoval Working , but the relatedEntity (ManyToOne , The Inverse side) getting a Null Value, as i want to only set deletedAt field to Null or Current Timestamp to identify Deleted or Not, The inverse relation Many2One assigning as Null by ORM Automatically,

Any way to avoid putting null on ManyToOne field??

as it putting null value, i cant restore relation by setting deletedAt to null, the relation Id has been gone,, 

is it bug ??

**Reply by totalard** [reply]
`[query] bulk = db.getCollection('product-variation').initializeUnorderedBulkOp({ session: undefined });bulk.find({ _id: ObjectId('628268b473e9167656686efc') }).update({ '$set': { deletedAt: ISODate('2022-05-16T15:08:44.020Z'), **product: undefined**, updatedAt: ISODate('2022-05-16T15:08:44.020Z') } });bulk.find({ _id: ObjectId('628268b473e9167656686efd') }).update({ '$set': { deletedAt: ISODate('2022-05-16T15:08:44.021Z'), **product: undefined**, updatedAt: ISODate('2022-05-16T15:08:44.021Z') } });bulk.find({ _id: ObjectId('628268b473e9167656686efe') }).update({ '$set': { deletedAt: ISODate('2022-05-16T15:08:44.021Z'), **product: undefined**, updatedAt: ISODate('2022-05-16T15:08:44.021Z') } });bulk.execute() [took 43 ms]`

this is the query executed while debugging , The bold fields are making these records cannot restored while restoring parent item, 

please help, how can i make it NOT NULL or NOT UNDEFINED in orphan removal

**Reply by B4nan** [reply]
Share some (minimal but complete) code so I can understand what exactly are you doing. 

Propagation cannot be currently disabled, but you can clean that up explicitly in the `onFlush` subscriber, removing the `null` value from change set's payload.

**Reply by totalard** [reply]
yeah, its okay, but i am using softdelete Subscriber on abstract Entity, so the abstract entity is inherited by many entities, so i have no idea how to find out in which orphanRemoval relation done this,

**Reply by totalard** [reply]
```
@Subscriber()
export class SoftDeleteSubscriber implements EventSubscriber {
  getSubscribedEntities(): EntityName<BaseSoftEntity>[] {
    return [BaseSoftEntity];
  }

  async onFlush(args: FlushEventArgs): Promise<void> {
    const changeSets = args.uow.getChangeSets();
    const toRemove = changeSets.filter(
      (cs) => cs.type === ChangeSetType.DELETE,
    );
    toRemove.forEach((cs) => {
      cs.type = ChangeSetType.UPDATE;
      cs.entity.deletedAt = new Date();
      cs.payload.deletedAt = new Date();
      // args.uow.recomputeSingleChangeSet(cs.entity); ///CHECK THIS
    });
  }
}
```
As i commented out this, now softdelete works fine, inverse relation not getting deleted too,

does this affect any trouble ??

**Reply by B4nan** [reply]
this should be fine, you dont need to recompute the changeset if you modify its payload yourself.

**Reply by totalard** [reply]
Can u suggest me , the procedure to do nested restore of orphanremoved entities in the softDelete.

Should I use entityMetadata to do it.

**Reply by jjhbw** [reply]
I had the same problem. Turned out the relevant `Collection`s were not initialised before deleting the root entity. Thanks for mentioning this! I think I overlooked the following sentence in the docs:

> This section is about application level cascading. For that to work, we need to have relations populated.

**Comment by nicolasb1607** [comment]
I had the same issue, and i found that you need to delete in cascade like this :

``` 
@OneToMany(() => OpinionEntity, (op) => op.user, {
    name: 'opinions',
    orphanRemoval: true,
  })
  opinions = new Collection<OpinionEntity>(this);
  ```
  
  and on the inverted side : 
  ```
  @ManyToOne(() => ArenaEntity, { deleteRule: 'cascade' })
  arena!: ArenaEntity;
  ```
  
  Hoping it will fix your issue
